### PR TITLE
#75 feat: search component

### PR DESCRIPTION
**Pull request by nafisreza**
## 📝 Description

Created a search component which is reusable throughout the website depending on the size, placeholder and callback function.

Relates to #75 

## 📸 Screenshots/Screen record (if applicable)

https://github.com/user-attachments/assets/29800976-ca86-49d2-8c97-3086a7d56fe5

## 🚀 Features
- Supports various sizes - ``xs``, ``sm``, ``md`` (default), ``lg``
- Has 3 different states - ``default``, ``hover``, ``focus``
- Search triggered ``onKeyDown`` - Enter
- Callback function ``onSearch`` is passed through the props to handle different cases

## ✅ Checklist

- [x] The component must be able to receive a callback function to handle the search action
- [x] The component must be able to receive a placeholder text
- [x] The component must have 4 size variants: xs, sm, md and lg
- [x] The component must have 3 states: default, hover and focus
- [x] The component must be documented with stories in Storybook

## 💡 How to Test
```
import Search from '@/components/ui/search'

<Search placeholder="Enter your search..." size="sm" onSearch={()=>alert("search triggered")} />
```

### CODE REVIEW

**Comment by mtbitcr** [comment]
Hi @nafisreza. Thanks for sticking to the PR form!

File: src/components/ui/search.tsx [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
+    if (event.key === "Enter") {
+      onSearch(searchFieldRef.current?.value || "");
+    }
+  };
+
+  return (
+    <div
+      onClick={focusSearchField}
+      className={`flex items-center gap-2 bg-elevation-50 border-[1px] border-divider-75 rounded-[8px] ${sizeClasses[size]} hover:bg-elevation-250 hover:border-divider-50 focus-within:bg-elevation-250 focus-within:border-divider-300 transition-all ease-in-out duration-200`}
```
Comment (cryptosalomao) [inline src/components/ui/search.tsx:35]: Use the `cn` helper from `@/lib/utils` to wrap these classes

File: src/components/ui/search.tsx [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  placeholder: string;
+  size?: "xs" | "sm" | "md" | "lg";
+}
+
+const sizeClasses = {
+  xs: "p-2 text-xs",
+  sm: "px-4 py-3 text-sm",
+  md: "p-4 text-sm",
+  lg: "px-4 py-5 text-sm",
+};
```
Comment (cryptosalomao) [inline src/components/ui/search.tsx:16]: Let's stick to the conventions for defining the variants with the `cva` utility from `class-variance-authority` package

**Review by cryptosalomao — CHANGES_REQUESTED** [review]
Hey @nafisreza great job! Overall, everything looks good. The only thing I want to ask you is to stick with the conventions for the variants definitions by using the existing patterns from the other components

Please check:
https://github.com/BitcreditProtocol/E-Bill-frontend/blob/main/src/components/ui/button.tsx

File: src/components/ui/search.tsx [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  const intl = useIntl();
+  const searchFieldRef = useRef<HTMLInputElement>(null);
+
+  const focusSearchField = () => {
+    searchFieldRef.current?.focus();
+  };
+
+  const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
+    if (event.key === "Enter") {
+      onSearch(searchFieldRef.current?.value || "");
```
Comment (cryptosalomao) [inline src/components/ui/search.tsx:28]: Only trigger the search if the value is not empty

File: src/components/ui/search.tsx [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+      />
+
+      <input
+        ref={searchFieldRef}
+        type="text"
+        placeholder={
+          placeholder ||
+          intl.formatMessage({
+            id: "contacts.search.placeholder",
+            defaultMessage: "Search for contacts...",
```
Comment (mjkeaton) [inline src/components/ui/search.tsx:57]:
```suggestion
            id: "common.search.placeholder",
            defaultMessage: "Search…",
```

Neutral terminology for the search component.

**Comment by mjkeaton** [comment]
Can this already be integrated somewhere to showcase (and enable visually inspecting the changes without manual adaptions)?

e.g. on the bills page:
```patch
diff --git a/src/pages/Bills.tsx b/src/pages/Bills.tsx
index 7da90a4..2cc4c3d 100644
--- a/src/pages/Bills.tsx
+++ b/src/pages/Bills.tsx
@@ -1,10 +1,12 @@
-import { FormattedMessage } from "react-intl";
+import { FormattedMessage, useIntl } from "react-intl";
 import { CalendarDaysIcon, ChevronsUpDownIcon, SearchIcon } from "lucide-react";
 import { Button } from "@/components/ui/button";
 import createBillIllustration from "@/assets/create-bill-illustration.svg";
 import { FormattedCurrency } from "@/components/FormattedCurrency";
+import Search from "@/components/ui/search";
 
 export function Bills() {
+  const intl = useIntl();
   return (
     <div className="flex flex-col min-h-fit h-screen gap-6 py-12 px-6 w-full">
       <div className="flex flex-col gap-3">
@@ -29,14 +31,11 @@ export function Bills() {
         </div>
 
         <div className="flex flex-col gap-2">
-          <div className="flex gap-2 items-center h-[46px] border-[1px] border-divider-75 rounded-[8px] px-4">
-            <SearchIcon size={16} strokeWidth={1} color="#1B0F00" />
-            <input
-              type="text"
-              placeholder="Search for bills..."
-              className="w-full bg-transparent text-sm font-medium text-text-300 outline-none"
-            />
-          </div>
+          <Search onSearch={() => {}} placeholder={intl.formatMessage({
+              id: "bills.search.placeholder",
+              defaultMessage: "Search for bills…",
+              description: "Placeholder text for search input",
+            })}/>
 
           <div className="flex gap-2">
             <Button variant="filter">
```

What do you think @nafisreza ? :pray:

**Comment by cryptosalomao** [comment]
> Can this already be integrated somewhere to showcase (and enable visually inspecting the changes without manual adaptions)?
> 
> e.g. on the bills page:
> 
> ```diff
> diff --git a/src/pages/Bills.tsx b/src/pages/Bills.tsx
> index 7da90a4..2cc4c3d 100644
> --- a/src/pages/Bills.tsx
> +++ b/src/pages/Bills.tsx
> @@ -1,10 +1,12 @@
> -import { FormattedMessage } from "react-intl";
> +import { FormattedMessage, useIntl } from "react-intl";
>  import { CalendarDaysIcon, ChevronsUpDownIcon, SearchIcon } from "lucide-react";
>  import { Button } from "@/components/ui/button";
>  import createBillIllustration from "@/assets/create-bill-illustration.svg";
>  import { FormattedCurrency } from "@/components/FormattedCurrency";
> +import Search from "@/components/ui/search";
>  
>  export function Bills() {
> +  const intl = useIntl();
>    return (
>      <div className="flex flex-col min-h-fit h-screen gap-6 py-12 px-6 w-full">
>        <div className="flex flex-col gap-3">
> @@ -29,14 +31,11 @@ export function Bills() {
>          </div>
>  
>          <div className="flex flex-col gap-2">
> -          <div className="flex gap-2 items-center h-[46px] border-[1px] border-divider-75 rounded-[8px] px-4">
> -            <SearchIcon size={16} strokeWidth={1} color="#1B0F00" />
> -            <input
> -              type="text"
> -              placeholder="Search for bills..."
> -              className="w-full bg-transparent text-sm font-medium text-text-300 outline-none"
> -            />
> -          </div>
> +          <Search onSearch={() => {}} placeholder={intl.formatMessage({
> +            id: "bills.search.placeholder",
> +            defaultMessage: "Search for bills…",
> +            description: "Placeholder text for search input",
> +          })} />
>  
>            <div className="flex gap-2">
>              <Button variant="filter">
> ```
> 
> What do you think @nafisreza ? 🙏
> 
> Edit: This can also be done in a follow-up PR!

@mjkeaton actually, it must be done on a story with Storybook framework. @nafisreza although you checked the checkbox for the Storybook requirement, it wasn't provided in the existing commits. Could you please include it? Let me know if you need help with it!

**Review by cryptosalomao — CHANGES_REQUESTED** [review]
Add story for this component

**Comment by nafisreza** [comment]
Alright guys, I will provide storybook components for this and the checkbox in upcoming commits.

**Review by mjkeaton — APPROVED** [review]

**Review by cryptosalomao — APPROVED** [review]

**Pull request merged**